File: salao-backend/src/main/java/com/salao/modules/categoria/CategoriaService.java
package com.salao.modules.categoria;

import com.salao.modules.produto.ProdutoRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
@RequiredArgsConstructor
public class CategoriaService {

    private final CategoriaRepository repository;
    private final ProdutoRepository produtoRepository;

    public List<Categoria> listar() {
        return repository.findAll();
    }

    public Categoria buscarPorId(Long id) {
        return repository.findById(id)
                .orElseThrow(() -> new RuntimeException("Categoria não encontrada"));
    }

    public Categoria salvar(CategoriaDTO dto) {
        if (repository.existsByNomeIgnoreCase(dto.nome()))
            throw new RuntimeException("Categoria já cadastrada");
        return repository.save(Categoria.builder().nome(dto.nome()).build());
    }

    public Categoria atualizar(Long id, CategoriaDTO dto) {
        Categoria cat = buscarPorId(id);
        cat.setNome(dto.nome());
        cat.setAtivo(dto.ativo());
        return repository.save(cat);
    }

    public void deletar(Long id) {
        buscarPorId(id);
        if (produtoRepository.existsByCategoriaId(id))
            throw new RuntimeException("Não é possível excluir: existem produtos vinculados a esta categoria");
        repository.deleteById(id);
    }
}

File: salao-frontend/src/services/categoriaService.ts
import axios from 'axios';

const API = 'http://localhost:8080/api';

export interface Categoria {
  id?: number;
  nome: string;
  ativo: boolean;
}

export const categoriaService = {
  listar: () => axios.get<Categoria[]>(`${API}/categorias`),
  buscar: (id: number) => axios.get<Categoria>(`${API}/categorias/${id}`),
  salvar: (data: Categoria) => axios.post<Categoria>(`${API}/categorias`, data),
  atualizar: (id: number, data: Categoria) => axios.put<Categoria>(`${API}/categorias/${id}`, data),
  deletar: (id: number) => axios.delete(`${API}/categorias/${id}`),
};